### ISSUE TITLE
Implement PSR-7 to replace HTTP Foundation throughout strategies

### ISSUE THREAD

**Issue by codeguy** [thephpleague/route#32]
What are your thoughts about supporting the PSR-7 request and response interfaces in the Request/Response strategy instead of a hard dependency on Symfony HTTP foundation?

**Comment by philipobenito** [comment]
100% the plan once it's accepted :+1:

**Comment by codeguy** [comment]
Sweet. Good news. Feel free to close this out.

**Comment by philipobenito** [comment]
Worth keeping open for clarification to others, I'll rename to reflect as a task

**Comment by philipobenito** [comment]
@codeguy would actually like your input re a middleware strategy if you'd be OK with me shooting you an email when I have a bit of time?

**Comment by codeguy** [comment]
Sounds great! You can reach me at hello at joshlockhart dot com.

**Comment by philipobenito** [comment]
Good man, thanks

**Comment by philipobenito** [comment]
Just another quick one, are you planning an implementation of PSR-7 for Slim 3? Obviously we'll be waiting a while for the Enterprises so if you are planning something, I'd give you a friendly nudge to make it a standalone package :-)

**Comment by codeguy** [comment]
Yes Slim 3 will have PSR-7 messages objects. I should probably split that into a separate component for sure.

**Comment by philipobenito** [comment]
Would be ideal for me, don't want to take the implementation on myself right now but obviously need to point to an implementation once I swap it out

**Comment by hannesvdvreken** [comment]
@philipobenito about middleware strategy: I think stackphp will depend on PSR-7 instead of http foundation once it's out. So imo following stackphp would be a good move.

Instead of depending on an implementation of PSR-7, league/route should depend on psr/http-message.

**Comment by philipobenito** [comment]
@hannesvdvreken PSR-7 === psr/http-message in this context, depending on the interfaces resulting from PSR-7. My point is that @codeguy is working on an implementation that I'd like to point to from a documentation perspective. Also I'm using Route heavily day to day so I'll need an implementation pretty quickly for my own use :-)

**Comment by codeguy** [comment]
@hannesvdvreken Yep, we all have the same goal. Everything will be coded for PSR-7 interfaces and not particular implementation of such.

**Comment by philipobenito** [comment]
That was always the aim with Route, ideally to write an application that was properly agnostic of it's framework/bootstrap, I want people to be able to move away from League\Route if they want to, and with the look of what @codeguy is doing with Slim 3, it would be minimal work to drop an app built on this router straight in to Slim and for it to work, first example of the FIG stuff working in the wild hopefully, which I'm kind of excited about.

**Comment by codeguy** [comment]
@philipobenito I'm pretty sure I'm going to use Route in Slim 3... if it doesn't support PSR7 by release date, I'll just write a quick strategy for it.

**Comment by philipobenito** [comment]
@codeguy that's fantastic news, if you need anything from me just shout, even if that be a specific tag for you to use until the PSR is official

**Comment by codeguy** [comment]
@philipobenito Pushed an experimental branch for you to review that integrates League/Route. You'll want to look at these files specifically:

* https://github.com/slimphp/Slim/blob/develop-route/Slim/Router.php
* https://github.com/slimphp/Slim/blob/develop-route/Slim/RouteStrategy.php
* https://github.com/slimphp/Slim/blob/develop-route/Slim/Route.php
* https://github.com/slimphp/Slim/blob/develop-route/Slim/App.php#L245

**Comment by hannesvdvreken** [comment]
@philipobenito: :ok-hand: pointing to an implementation from docs. We're on the same page then ;-)

**Comment by philipobenito** [comment]
All looks good @codeguy, I wrestled with the idea of catching not found etc within the strategy but settled on the opinion that it's better outside of there for the user to decide what to do. Will you be exposing the container or continuing with Pimple like it says in your blog post?

**Comment by codeguy** [comment]
May as well move over to League/container to reduce dependencies.

**Comment by philipobenito** [comment]
Ok so I'm planning a simplification of the API which will probably be much better for you, we can discuss what changes I have planned before release. Do you have a release date you're aiming for then I can aim to get v2 of both packages ready for then? Route won't be changing much at all, just the dependency change and a possible filesystem caching layer for defined routes so that it's not interacted with every request (this is something built in to FastRoute). Container again isn't gonna be changing in terms of functionality, just a big internal tidy up and improving the cohesiveness between features.

**Comment by codeguy** [comment]
No hard release date yet.

**Comment by philipobenito** [comment]
Ok, well I'll be starting my planned works the next couple of days and shouldn't take me too long so I'll keep you updated to public changes.

**Comment by codeguy** [comment]
:+1:

**Comment by codeguy** [comment]
@philipobenito Not sure if your changes will include this or not, but can you create a new tag/branch that does not have the Symfony HTTP Foundation dependency in `composer.json`? No immediate rush.

**Comment by philipobenito** [comment]
Yeah, I'll create you a branch later tonight and alias a packagist tag to it then any specifics for slim can be put in there

**Comment by philipobenito** [comment]
@codeguy hopefully this will help for now, http exceptions and and json responses will be broken for now but will fix that tomorrow https://github.com/thephpleague/route/commit/1f97cd57f3350517e5980328c911e6751b0274f0

**Comment by codeguy** [comment]
@philipobenito After some experimentation, I've found it's going to be easier and cleaner to use FastRoute directly. However, I'm definitely going to extract the PSR-7 stuff into a separate component and do what I can to advertise and help league/route.

**Comment by philipobenito** [comment]
@codeguy That makes sense based on what you're doing, look forward to the PSR-7 package though, aiming to get that implemented asap.

**Comment by sagikazarmark** [comment]
@codeguy @philipobenito are you sure that stackphp will rely on PSR-7? Because I heard other news. I was also thinking about creating a middleware package for PSR-7, because [phly/conduit](https://github.com/phly/conduit) is not really what I imagine as the clone of stackphp.

In the meantime check [phly/http](https://github.com/phly/http) which is a HTTP Message implementation created by the editor of the PSR.

**Comment by sagikazarmark** [comment]
Some update: PSR 7 is out. There are already a few implementations out there, the most sane ones:

- [phly/http](https://github.com/phly/http)
- [guzzle/psr7](https://github.com/guzzle/psr7)

They both provide the `psr/http-message-implementation` virtual package.

**Comment by philipobenito** [comment]
@sagikazarmark v2 is due soon and will depend on the abstractions but not provide an implementation  by default. It will simply suggest several alternatives and the documentation/skeleton applications will handle the implementation details.

**Comment by sagikazarmark** [comment]
The provided virtual package is great for both directly requiring or suggesting.

**Comment by sagikazarmark** [comment]
I would probably extract the PSR7 part from the main package and move it to a separate one so that you can require the virtual package instead of suggesting it.

**Comment by ojhaujjwal** [comment]
I agree with @sagikazarmark about moving PSR7 implementation to a separate package!

**Comment by shadowhand** [comment]
@philipobenito do you want me to work on a PR for this?

**Comment by philipobenito** [comment]
It's actually already being worked on for v2, was waiting for the vote to pass before finalising but thank you.

**Comment by shadowhand** [comment]
@philipobenito when can we expect to see a branch?

**Comment by dolfelt** [comment]
@philipobenito Which implementations are being suggested going forward using PSR-7 with Route?

**Comment by philipobenito** [comment]
It will be completely up to the user which implementation is used but there will be a section in the docs to outline options.

**Comment by philipobenito** [comment]
Next 1-2 weeks

**Comment by sagikazarmark** [comment]
It seems that the complete replace of Symfony HttpFoundation might not be necessary:

http://symfony.com/blog/psr-7-support-in-symfony-is-here

To preserve compatibility (for example with Stack based applications, like Proton) this compatibility layer could be used.

For the record I suggest preserving compatibility next to the new, PSR-7 strategy which should be the recommended one.

**Comment by philipobenito** [comment]
This has already been discussed on several threads and is being implemented for v2.

**Comment by sagikazarmark** [comment]
Sorry, I didn't find any reference to symfony's compatibility layer. Also, it just have been released.

**Comment by philipobenito** [comment]
Less Symfony specific but the fact that strategies are going to be split out and HTTP Foundation will still be supported.

**Comment by alexbilbie** [comment]
Hey @philipobenito where are you with this? Am happy to help with this

**Comment by philipobenito** [comment]
Hey man. Already done, just got some changes to make to Container before I release, next week some time for both v2 releases.



—

**Comment by alexbilbie** [comment]
:heart:

**Comment by mwillbanks** [comment]
@philipobenito any chance of creating a branch with some of this so that it can be exposed and some of us can start testing it?

**Comment by sagikazarmark** [comment]
:+1: I am also curious about new features.

**Comment by hannesvdvreken** [comment]
+1
